### PR TITLE
[Bug] [seatunnel-api-flink] Use the constructor instead of setting the pipeline.jars

### LINKED ISSUE #2975: [Discuss][Core] How to up-compatible flink at api level
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.

### What happened

this is a subtask of #2927 
Now,Later versions of FLink do not allow direct set up `Configuration`,the `Configuration` is unmodifiable,and i think constructors are more stable,so i want to set the configuration via the constructor.

### SeaTunnel Version

dev

### SeaTunnel Config

```conf
-
```

### Running Command

```shell
-
```

### Error Exception

```log
java.lang.UnsupportedOperationException: The configuration is unmodifiable; its contents cannot be changed.
```

### Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by liugddx**
close #2975

## Purpose of this pull request

Now,Later versions of FLink do not allow direct set up Configuration,the Configuration is unmodifiable,and i think constructors are more stable,so i want to set the configuration via the constructor.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by liugddx** [comment]
@CalvinKirs @EricJoy2048 @ashulin @Hisoka-X This pr is a subtask for Upgrade flink. For details in #2975

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 1]
```diff
@@ -50,15 +57,29 @@ public class FlinkExecution implements TaskExecution {
...
-        this.transformPluginExecuteProcessor = new TransformExecuteProcessor(flinkEnvironment, jobContext, config.getConfigList(Constants.TRANSFORM));
-        this.sinkPluginExecuteProcessor = new SinkExecuteProcessor(flinkEnvironment, jobContext, config.getConfigList(Constants.SINK));
+        JobContext jobContext = new JobContext();
+        jobContext.setJobMode(new FlinkEnvironmentFactory(config).getJobMode(config.getConfig("env")));
+
+        this.sourcePluginExecuteProcessor = new SourceExecuteProcessor(jarPaths, config.getConfigList(Constants.SOURCE), jobContext);
+        this.transformPluginExecuteProcessor = new TransformExecuteProcessor(jarPaths, config.getConfigList(Constants.TRANSFORM), jobContext);
+        this.sinkPluginExecuteProcessor = new SinkExecuteProcessor(jarPaths, config.getConfigList(Constants.SINK), jobContext);
+
+        this.flinkEnvironment = new FlinkEnvironmentFactory(this.registerPlugin(config, jarPaths)).getEnvironment();
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java:78]: Don't create `FlinkEnvironmentFactory` twice

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 2]
```diff
@@ -90,6 +111,33 @@ private void registerPlugin() {
 
         pluginsJarDependencies.forEach(url -> FlinkCommon.ADD_URL_TO_CLASSLOADER.accept(Thread.currentThread().getContextClassLoader(), url));
 
-        flinkEnvironment.registerPlugin(pluginsJarDependencies);
+        jarPaths.addAll(pluginsJarDependencies);
+    }
+
+    private Config registerPlugin(Config config, List<URL> jars) {
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java:118]: Maybe should named `injectJarsToConfig`

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 3]
```diff
@@ -90,6 +111,33 @@ private void registerPlugin() {
...
+        config = this.parseConfig(config, ConfigUtil.joinPath("env", "pipeline", "jars"), jars);
+        return this.parseConfig(config, ConfigUtil.joinPath("env", "pipeline", "classpaths"), jars);
+    }
+
+    private Config parseConfig(Config config, String path, List<URL> jars) {
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: Should change name too, Can't know it want to do what if just read name.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 4]
```diff
@@ -90,6 +111,33 @@ private void registerPlugin() {
 
         pluginsJarDependencies.forEach(url -> FlinkCommon.ADD_URL_TO_CLASSLOADER.accept(Thread.currentThread().getContextClassLoader(), url));
 
-        flinkEnvironment.registerPlugin(pluginsJarDependencies);
+        jarPaths.addAll(pluginsJarDependencies);
+    }
+
+    private Config registerPlugin(Config config, List<URL> jars) {
+        config = this.parseConfig(config, ConfigUtil.joinPath("env", "pipeline", "jars"), jars);
+        return this.parseConfig(config, ConfigUtil.joinPath("env", "pipeline", "classpaths"), jars);
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: Change the doc to tell user the new way add jar. Also this function should support spark later.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 5]
```diff
@@ -90,6 +111,33 @@ private void registerPlugin() {
...
+    }
+
+    private Config parseConfig(Config config, String path, List<URL> jars) {
+
+        if (config.hasPath(path)) {
+            List<URL> paths = Arrays.stream(config.getString(path).split(";")).map(uri -> {
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: Use `Set` to aovid repeated jar path.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java [hunk 6]
```diff
@@ -50,15 +57,29 @@ public class FlinkExecution implements TaskExecution {
...
     private final PluginExecuteProcessor transformPluginExecuteProcessor;
     private final PluginExecuteProcessor sinkPluginExecuteProcessor;
+    private final List<URL> jarPaths;
 
     public FlinkExecution(Config config) {
-        this.flinkEnvironment = new FlinkEnvironmentFactory(config).getEnvironment();
-        JobContext jobContext = new JobContext();
-        jobContext.setJobMode(flinkEnvironment.getJobMode());
+        try {
+            jarPaths = new ArrayList<>(Collections.singletonList(
```
Comment (Hisoka-X) [inline seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java:65]: Use `Set` to aovid repeated jar path.

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java (hunk 4 above)
Comment (liugddx) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: > Change the doc to tell user the new way add jar. Also this function should support spark later.

This change does not require a change in usage,it's just the way the jar is injected that has changed.

Comment (Hisoka-X) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: But you support add jar use `env.pipeline.jars` in config

Comment (liugddx) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: > Change the doc to tell user the new way add jar. Also this function should support spark later.
Spark uses the command line,i don't think it needs changing.

**Comment by liugddx** [comment]
Done. @Hisoka-X

File: seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java (hunk 4 above)
Comment (liugddx) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: > But you support add jar use `env.pipeline.jars` in config
This is only used by flink,does spark need this way?

Comment (liugddx) [reply seatunnel-core/seatunnel-flink-starter/src/main/java/org/apache/seatunnel/core/starter/flink/execution/FlinkExecution.java]: > But you support add jar use `env.pipeline.jars` in config

I see.

**Review by Hisoka-X — APPROVED** [review]
Great! LGTM. Waiting CI Pass.

**Review by EricJoy2048 — APPROVED** [review]
LGTM

**Comment by Hisoka-X** [comment]
Retry CI after #3183 merged

**Comment by liugddx** [comment]
There seems to be a problem, too



&nbsp;




------------------&nbsp;Original&nbsp;------------------
From:                                                                                                                        "apache/incubator-seatunnel"                                                                                    ***@***.***&gt;;
Date:&nbsp;Tue, Oct 25, 2022 08:54 PM
***@***.***&gt;;
***@***.******@***.***&gt;;
Subject:&nbsp;Re: [apache/incubator-seatunnel] [Bug] [seatunnel-api-flink] Use the constructor instead of setting the pipeline.jars (PR #2982)





 
Retry CI after #3183 merged

**Comment by Hisoka-X** [comment]
> There seems to be a problem, too &nbsp;
> […](#)
> ------------------&nbsp;Original&nbsp;------------------ From: "apache/incubator-seatunnel" ***@***.***&gt;; Date:&nbsp;Tue, Oct 25, 2022 08:54 PM ***@***.***&gt;; ***@***.******@***.***&gt;; Subject:&nbsp;Re: [apache/incubator-seatunnel] [Bug] [seatunnel-api-flink] Use the constructor instead of setting the pipeline.jars (PR #2982) Retry CI after #3183 merged — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

We will fix it.

**Pull request merged**